### PR TITLE
feat(CLDX-78): add windows binary signing step

### PR DESCRIPTION
**Pull request by swickersh**
This step is part of multiple commits for the sign-binaries task. Specifically, this is the Windows signing step. The rest of the task and steps will be implemented with CLDX Jiras as marked in the code comments.

Currently, no pipelines call this task yet, so the TODO code will not cause any failures as this time.

### CODE REVIEW

File: tasks/sign-binaries/sign-binaries.yaml [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+metadata:
+  name: sign-binaries
+  labels:
+    app.kubernetes.io/version: "0.2.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: release
+spec:
+  description: >-
+    Tekton task to sign mac and windows binaries using system specific signing hosts
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: A small detail, but there's probably no reason to have different descriptions in the README and here.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  params:
+    - name: quayURL
+      type: string
+      description: quay URL of the repo where content will be shared
+      default: quay.io/konflux-artifacts
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: README says there is no default. But this doesn't look like a valid value anyway, right? Description says it should be a repo, so should be a longer path.

But seeing now that you're adding `/unsigned` to it, maybe it's meant as more of a prefix?

File: tasks/sign-binaries/sign-binaries.yaml [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      description: Workspace to save the results to
+  results:
+    - name: unsignedWindowsDigest
+      type: string
+      description: |
+        The digest of the unsigned content pushed using ORAS for signing hosts
+    - name: signedWindowsDigest
+      type: string
+      description: |
+        The digest of the unsigned content pushed using ORAS for signing hosts
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: This seems wrong. Should be "signed content" that the hosts signed, right?

File: tasks/sign-binaries/sign-binaries.yaml [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  steps:
+    - name: push-unsigned-using-oras
+      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
+      script: |
+        #!/bin/bash
+
+        # TODO CLDX-134
+
+        output=$(oras push "$(params.quayURL)/unsigned" .)
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml:55]: Maybe also print the output? It could help with debugging if something unexpected happens.

Also, what's the `.` ? Does it mean "push everything in the current directory"? But this will likely be empty. Will this be covered in another PR? It doesn't seem to be part of CLDX-134 .

File: tasks/sign-binaries/sign-binaries.yaml [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    - name: windowsCredentials
+      type: string
+      description: Secret to interact with the Windows signing host
+    - name: pipelineRunUid
+      type: string
+      description: Unique identifier for the pipeline run
+  volumes:
+    - name: windows-ssh-key-vol
+      secret:
+        secretName: windows-ssh-key
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: At first I didn't realize there would be two windows secret and I thought this should use the parameter. I guess it probably makes sense to store the ssh private key separately, because it will likely always be the same. But I wonder if it would make sense to also use a task parameter for this. Even if there is a default value that is used in most cases, I think it's still probably nicer than hardcoding it here. WDYT?

File: tasks/sign-binaries/sign-binaries.yaml [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        The digest of the unsigned content pushed using ORAS for signing hosts
+    - name: signedWindowsDigest
+      type: string
+      description: |
+        The digest of the unsigned content pushed using ORAS for signing hosts
+  steps:
+    - name: push-unsigned-using-oras
+      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
+      script: |
+        #!/bin/bash
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Nit: Maybe use `#!/usr/bin/env bash` to be consistent with other tasks? But if you prefer this, it's fine as well.

Also, maybe add `set -eux` or at least `set -x`? It usually doesn't hurt to have more details in the log.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        - name: QUAY_PASS
+          valueFrom:
+            secretKeyRef:
+              name: $(params.quaySecret)
+              key: password
+      script: |
+        #!/bin/bash
+        mkdir -p ~/.ssh
+        cp "/etc/secrets/id_rsa" ~/.ssh/id_rsa
+        chmod 600 ~/.ssh/id_rsa
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: I think the `.ssh` dir also needs a chmod - it needs to have 700.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        unsigned_digest=$(cat "$(results.unsignedWindowsDigest.path)")
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Are you sure this will work? When I searched for how to pass data between steps, I found the recommended way is to use an empty dir. But if this works for you when I guess it's fine too.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        # EOF should be unquoted here to allow for variable expansion on the client side
+        # shellcheck disable=SC2086,SC2087
+        ssh -i ~/.ssh/id_rsa -o StrictHostKeyChecking=no -p "${WINDOWS_PORT}" "${WINDOWS_USER}@${WINDOWS_HOST}" << EOF
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Maybe it's ok for now, but it would be great if we could have a fingerprint of the windows host and inject it into `~/.ssh/known_hosts` and then the host checking could happen as normal.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+        set SIGNTOOL="C:\Program Files (x86)\Windows Kits\10\bin\10.0.20348.0\x64\signtool.exe"
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Isn't there a risk that this specific version won't be there? Just asking.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        set ORAS="C:\Program Files\Common Files\bin\oras.exe"
+
+        mkdir %TEMP%\$(params.pipelineRunUid) && cd /d %TEMP%\$(params.pipelineRunUid)
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml:117]: I wonder if the backslash might confuse Tekton and prevent it from substituting parameter value. But I guess sooner or later you will find out.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+          exit /B %ERRORLEVEL%
+        )
+
+        echo [%DATE% %TIME%] Signing of Windows binaries completed successfully
+
+        %ORAS% push $(params.quayURL)/signed:$(params.pipelineRunUid) . > oras_push_output.txt 2>&1
+        for /f "tokens=2,3 delims=: " %a in ('findstr "Digest:" oras_push_output.txt') do @echo %a:%b > digest.txt
+
+        EOF
+        DIGEST=$(ssh -i ~/.ssh/id_rsa -o StrictHostKeyChecking=no -p "${WINDOWS_PORT}" \
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: I wonder if potentially the ssh server won't print some greetings or something at the top. Wouldn't it be better to just scp the file?

File: tasks/sign-binaries/tests/mocks.sh [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+count_file="/tmp/ssh_count.txt"
+if [[ ! -f "$count_file" ]]; then
+    echo "ssh_call_count=0" > "$count_file"
+fi
+
+
+function ssh() {
+    # Read the current ssh_call_count from the file
+    ssh_call_count=$(grep -oP '(?<=ssh_call_count=)\d+' "$count_file")
```
Comment (mmalina) [inline tasks/sign-binaries/tests/mocks.sh]: Wouldn't it be easier to just have the number in the file and nothing else? Just saying :)

File: tasks/sign-binaries/sign-binaries.yaml (hunk 1 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Both match now. Thanks

File: tasks/sign-binaries/sign-binaries.yaml (hunk 2 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: This was a remnant from testing that I forgot to remove. No default, thanks!

File: tasks/sign-binaries/sign-binaries.yaml (hunk 5 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Yep that would be more modular for sure. Parameterized it but kept default. Thanks

File: tasks/sign-binaries/sign-binaries.yaml (hunk 3 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: copy/paste error. Thanks for catching

File: tasks/sign-binaries/sign-binaries.yaml (hunk 6 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Added `#!/usr/bin/env bash` for consistency.
I originally was using `-x` (xtrace) but the Quay credentials and windows host/port were showing up in the logs which I didn't want. 

I put `set -eu` so we'll exit on error or if unset var is encountered. Would that be sufficient?

File: tasks/sign-binaries/sign-binaries.yaml (hunk 4 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml:55]: I only added this small snippet to make mocking the Windows step easier and more realistic. This step will be enhanced quite a bit with CLDX-134. (I just edited the Jira description to clarify this)
Pkhander already wrote the step and will be submitting a PR soon.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 7 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Thanks, corrected

File: tasks/sign-binaries/sign-binaries.yaml (hunk 8 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Yes I tested this.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 9 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Done. Thanks

File: tasks/sign-binaries/sign-binaries.yaml (hunk 11 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml:117]: It worked in testing... but maybe I should reconsider and SCP the file

File: tasks/sign-binaries/tests/mocks.sh (hunk 13 above)
Comment (swickersh) [reply tasks/sign-binaries/tests/mocks.sh]: :-) yes. That was a remnant from when I was using a single file for multiple tests scenarios. Thanks!

File: tasks/sign-binaries/sign-binaries.yaml (hunk 10 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: I was unsuccessful adding signtool and oras to the path on windows. I've reached out to someone more windows savvy than me for suggestions. Will update if we can get it sorted out. Thanks

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Pushing other changes up. Keep this comment unresolved and I'll test SCP'ing the file over in the meantime. Thanks!

File: tasks/sign-binaries/sign-binaries.yaml (hunk 3 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: NP. But I think it's still slightly wrong - the unsigned one is for the hosts to pull, but this one is actually pushed by the hosts for the task to pull afterwards.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 4 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml:55]: OK, sounds good, thanks.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 6 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: What we started doing in other places is start with `-x` and then for the snippet where some secret would be leaked, you add `set +x` above it and then `set -x` below. That way you still have more logging most of the time, but avoid leaking anything. WDYT of that?

File: tasks/sign-binaries/sign-binaries.yaml (hunk 10 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: OK, cool. Leaving this open for now.

File: tasks/sign-binaries/tests/mocks.sh (hunk 13 above)
Comment (mmalina) [reply tasks/sign-binaries/tests/mocks.sh]: ok, cool :)

File: tasks/sign-binaries/sign-binaries.yaml (hunk 3 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: The workflow is as follows:
push `unsigned` binaries from konflux space to registry with oras > `unsigned digest`
pass `unsigned digest` to windows host so it pulls the `unsigned` binaries 
windows signs the binaries and pushes `signed` binaries with oras back to registry > `signed digest`
`signed digest` used by next step to pull signed binaries to konflux space

Does that make sense? the registry is only used as a means to transfer the signed/unsigned binaries between Konflux and the signing hosts. The final signed binaries are pushed to the RH Developer Portal with CDN and CGW tasks.

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: Yeah, ok, so maybe I'm just misunderstanding the language in the descriptions then? 
"The digest of the unsigned content pushed using ORAS for signing hosts" - I read that as "The digest of the unsigned content pushed using ORAS for signing hosts [to consume]". Or is that not how it's meant? If it is, then for the signed one it sounds wrong, no?

"The digest of the signed content pushed using ORAS for signing hosts [to consume]" - this one is actually pushed by the signing hosts for the rest of the pipeline to consume.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 10 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Paths are set on Windows host now.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Now SCP'ing the file over. Good suggestion thanks

File: tasks/sign-binaries/sign-binaries.yaml (hunk 11 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml:117]: File transferred with SCP now

File: tasks/sign-binaries/sign-binaries.yaml (hunk 3 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: I think I see the confusion now. I've reworded it to hopefully make it more clear. Thanks!

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: Yep, much better now, thanks!

File: tasks/sign-binaries/sign-binaries.yaml [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        > oras_push_output.txt 2>&1
+
+        for /f "tokens=2,3 delims=: " %%a in ('findstr "Digest:" oras_push_output.txt') do @echo %%a:%%b > digest.txt
+        EOF
+        set -x
+        scp "$SCP_OPTS" "$signing_script_file" \
+        "${WINDOWS_USER}@${WINDOWS_HOST}:C:/Users/Administrator/AppData/Local/Temp/signing_script.bat"
+
+        # Execute the script on the Windows host
+        ssh "$SSH_OPTS" "C:/Users/Administrator/AppData/Local/Temp/signing_script.bat"
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml:150]: Don't you need to set it to executable? I haven't touched Windows in a long time. Maybe it's enough that it has the .bat suffix?

File: tasks/sign-binaries/sign-binaries.yaml (hunk 11 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml:117]: Cool!

File: tasks/sign-binaries/sign-binaries.yaml [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+        # disable shellcheck for escaping the pipelineRunUid as we want that evaulated on client side
```
Comment (mmalina) [inline tasks/sign-binaries/sign-binaries.yaml]: Typo: evaulated -> evaluated

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: I see scp is used to copy over the signing script to the Windows host. But this was about copying the digest back and I still see this being done via ssh + type.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 14 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml:150]: Nope, tested it and it works as expected.

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml:150]: Cool, thanks.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 15 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: thanks, fixed

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Ah, yea might as well use SCP that direction too. Thanks done.

File: tasks/sign-binaries/sign-binaries.yaml [hunk 16]
```diff
@@ -0,0 +1,172 @@
+---
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: sign-binaries
+  labels:
+    app.kubernetes.io/version: "0.2.1"
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml]: This should be 0.1.0

File: tasks/sign-binaries/sign-binaries.yaml [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        signtool verify /v /pa %TEMP%\$(params.pipelineRunUid)\*
+
+        if errorlevel 1 (
+          echo Verification of binaries failed
+          exit /B %ERRORLEVEL%
+        )
+
+        echo [%DATE% %TIME%] Signing of Windows binaries completed successfully
+
+        oras push $(params.quayURL)/signed:$(params.pipelineRunUid) %TEMP%\$(params.pipelineRunUid) \
```
Comment (johnbieren) [inline tasks/sign-binaries/sign-binaries.yaml:140]: I think you'd need the select-oci-auth piece before this, right @mmalina ? Like https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-pyxis-image/create-pyxis-image.yaml#L118

File: tasks/sign-binaries/sign-binaries.yaml (hunk 16 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: done. thanks!

File: tasks/sign-binaries/sign-binaries.yaml (hunk 17 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml:140]: I was doing `oras login` on line 119 with  a user:pass. 
Are you suggesting I use a config file with credentials instead? I'm not familiar with `select-oci-auth` and couldn't trace where it originates from that url.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: (Sorry for the delay, I was out.)

Thanks. But this brings another question: Where does the binary come from originally? That part is not done yet, but I assume the binary will first be fetched from somewhere and then the task will push it using oras. But now that we know we can use scp, wouldn't it make more sense to simply copy the binary using scp? Isn't using oras pull/push adding an unnecessary overhead? Or are there any benefits that I'm missing?

File: tasks/sign-binaries/sign-binaries.yaml (hunk 17 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml:140]: `select-oci-auth` originates from here: https://github.com/konflux-ci/oras-container/blob/main/hack/select-oci-auth.sh

If you're using a user+pass directly, then it doesn't apply. But it's a good question how we want to approach this. Because one way would be to rely on pull secrets linked to the service account - that's how all of our other pulls and pushes work - we don't do anything special (other than the `select-oci-auth` hack) and it just works, because those secrets are there. Hmm, but then you also need this working on the windows host. In theory you could set up a docker config there beforehand. But maybe it makes more sense to just use user+pass in this case.

(But note that I'm asking if we couldn't simply scp the files in another thread.)

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml]: Fair question.  This was covered in the [design doc](https://docs.google.com/document/d/1FmJmOinT6BSkJiK_AWK3PFDBF_08UqSQUakkivKhKAk/edit?usp=sharing) and approved by Ralph, Brian and others.  

We considered SCP, NFS share and other methods for sharing content. 

If you have questions that aren't covered in the doc or the linked Slack thread from the doc can you ping me on Slack? That might be more efficient and I can summarize our conversation in a comment here afterwards for transparency.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 17 above)
Comment (swickersh) [reply tasks/sign-binaries/sign-binaries.yaml:140]: Thanks for sharing. I don't believe `select-oci-auth` applies here.

Instead of Quay image-pull-secrets we have a dedicated Quay org and robot/service account. The credentials are passed to Windows/mac hosts as part of the task during the SCP transfer.

Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml:140]: Yep, I agree, it should be fine.

File: tasks/sign-binaries/sign-binaries.yaml (hunk 12 above)
Comment (mmalina) [reply tasks/sign-binaries/sign-binaries.yaml]: Thanks for the link. I sent you a PM with more details, but alright, let's go with oras pull/push then.

**Review by mmalina — APPROVED** [review]

**Review by johnbieren — APPROVED** [review]

**Comment by mmalina** [comment]
Ignoring e2e here, since this task is not even included in any pipeline yet.
